### DISCUSSION TITLE
Mount additional disks

### DISCUSSION THREAD

**Discussion by thirdeyenick** [gokrazy/gokrazy#235] [Q&A] [answered]
Hi there. I wanted to ask if there is already a possibility to mount additional USB connected disks in gokrazy? I'd like to use minio with additional USB disks on my gokrazy powered raspberry pi. If this is not possible so far, I was thinking about an '/etc/fstab' similar approach, but instead of a file, the additional disks can be specified in the gokrazy config. Then the supervise init process could take care of mounting before starting all configured applications and unmounting after all the applications have finished.

**Comment by stapelberg** [comment, marked as answer]
This is not currently implemented.

You can deploy a small Go program to mount the external disk, though. Hopefully the minio process recovers automatically? Otherwise, you could disable starting minio automatically and start it from your program.

**Reply by thirdeyenick** [reply]
Thank you for the quick response. Yeah, I thought about this as well and was about to start to write such a wrapper which first mounts everything and then starts minio. But then I thought, what if 2 different apps need the same disk/partition mounted (they just write to different sub-directories on that mount)? Sure, I could then start both apps from the same wrapper :-) But the point is that mounting and providing a "mount dependency" is easier provided from a central point which also knows when to safely unmount all additional disks. Nevertheless, writing a wrapper is also possible :-)

**Reply by stapelberg** [reply]
Filed https://github.com/gokrazy/gokrazy/issues/236 to track this feature

**Reply by thirdeyenick** [reply]
Thank you very much for creating the issue. I got a bit further with this and my current plan is to provide a custom [Boot function](https://github.com/gokrazy/gokrazy/blob/main/gokrazy.go#L218) by using a replace directive in my `builddir/github.com/gokrazy/gokrazy/go.mod` file. That custom build function then checks for the existence of a specific `mounts.json` file on `/perm`. It then reads the contents and tries to mount the devices. That way I don't need to have more config files in the generated gokrazy disk image, but I can create the file myself on the `/perm` partition. This doesn't really fit the 'mount external devices to /perm' use case of the created issue though. 

It also seems I don't need to umount the disks at the end (at least I couldn't find any umount calls in the current existing gokrazy code so far).

I thought about identifying the device file properly and I think I will go with the PARTUUID of the partition to mount. This should be existing for MBR and GPT formatted disks. For this I want to use [go-diskfs](https://github.com/diskfs/go-diskfs) where my corresponding [PR](https://github.com/diskfs/go-diskfs/pull/198) got merged. As the filesystem needs to be specified in the `mounts.json` file I hope I don't need to probe for filesystems this way.

Lets see if all of this works out in the end :-)

**Reply by flber** [reply]
@stapelberg how would you recommend going about writing a wrapper to mount an external drive if (like you mention in #236) there aren't fixed symlinks to device nodes? More specifically, how do you find the location of a given external drive in the filesystem to then mount with syscalls? I'm trying to set up a Jellyfin box, but this has been stymieing me for a little while now :P

[figured that out, the device was actually in `/dev/` I guess - `blkid` helped]

Are exfat devices supported? From the contents of `/proc/filesystems` it doesn't seem like it is, and running `mount /dev/sda1 /perm/mnt/` returns `... failed: Invalid argument`, so how might I patch that in?

**Reply by stapelberg** [reply]
Looks like Linux 5.4+ has a driver for exfat, so this might be as simple as setting `CONFIG_EXFAT_FS=y` in gokrazy/kernel/cmd/gokr-build-kernel/config.addendum.txt. See the https://github.com/gokrazy/kernel README for details.

**Reply by flber** [reply]
Perfect, that seems to work! - would that be something to make a pr for? 
Jellyfin had some trouble reading from the drive, but some futzing and it works now :)

**Reply by damdo** [reply]
@flber May be worth writing a little guide about this on the gokrazy website :)
It might be useful to others!

**Reply by flber** [reply]
@damdo yeah that could definitely be good, unfortunately it seems my setup is a little fragile right now - I managed to fully crash my install :P (I think it was jellyfin's fault).
But the exfat thing would probably be helpful to document even if it's not part of a larger example.

**Reply by stapelberg** [reply]
Yeah, feel free to send a PR to include exfat in gokrazy/kernel. If it increases the kernel size significantly we should enable it as a module.

**Reply by flber** [reply]
Okay, a [PR](https://github.com/gokrazy/kernel/pull/562) has been made

**Comment by thirdeyenick** [comment]
For the time being, I [added a rough implementation](https://github.com/thirdeyenick/gokrazy/commit/2a01b4976736d028c0e1ab1552b34b132aabddd8) which allows to mount additional partitions, by specifying them in a `/perm/mount-disks.json` file:

```json
{
  "disks": [
    {
      "partUUID": "<found via blkid>",
      "type": "ext4",
      "mountpoint": "/perm/mounts/disk1",
      "options": ""
    }
  ]
}
```

I then used a `replace` directive in my `<instance>/builddir/github.com/gokrazy/gokrazy/go.mod` file to point to it. So far it works quite well for my use case, but a more advanced implementation might be needed for this.

**Reply by stapelberg** [reply]
Thanks, I referenced this over in issue #236
